### ISSUE TITLE
Leverage new Lock types

### ISSUE THREAD

**Issue by jamescrosswell** [getsentry/sentry-dotnet#3689]
C# 13 includes a [new Lock type](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-13#new-lock-object) intended to be used with the lock keyword.

Using Lock types as lock objects, the compiler can perform some optimisations that aren't possible when using standard object lock objects. It's a small thing, but we should probably replace the lock objects in our SDK to enable these optimisations.

**Comment by Flash0ver** [comment]
[Polyfill v6.12.0](https://github.com/SimonCropp/Polyfill/milestone/101) adds the `System.Threading.Lock` type.
It satisfies the shape for the "C# 13 lock statement".
The implementation uses `System.Threading.Monitor` under the hood.

I am not 100 % certain that this is 100 % correct, because there is a difference in IL-code-gen between the two `lock`-variants:
```diff
-object obj = new object();
-bool lockTaken = false;
+Lock @lock = new Lock();
+Lock.Scope scope = @lock.EnterScope();
try
{
- Monitor.Enter(obj, ref lockTaken);
  Console.WriteLine("mutual-exclusion");
}
finally
{
- if (lockTaken)
- {
-   Monitor.Exit(obj2);
- }
+ scope.Dispose();
}
```

where - with the polyfilled version on `<= net8.0` - the implemented `Monitor.Enter` called via `EnterScope` would then be outside of the `try`-block.

Could this cause problems for our use cases?

Alternatively, we could `#if NET9_0_OR_GREATER` all usages of `System.Threading.Lock`, and use `object` instead, but that could bloat the source code quite a bit.
Alternatively, we could alias `using Lock = object;` globally (e.g. via C# or MSBuild) for `#if !NET9_0_OR_GREATER`.

Just some thoughts for the implementer.

**Comment by jamescrosswell** [comment]
> Alternatively, we could #if NET9_0_OR_GREATER all usages of System.Threading.Lock, and use object instead, but that could bloat the source code quite a bit.

Yeah that's why we haven't done this already. That seemed like a pretty ugly solution (especially since we have no prospect of ever being able to drop net472 or netstandard). 

I think the Polyfill implementation of Lock is the best option. The IL will never be the same as the net9.0 version but that's OK as long as we don't take a step backwards, performance wise, over what we have now.

**Comment by Flash0ver** [comment]
After a discussion
- we're not sure about the Polyfill not causing potential issues on .NET Framework (or Runtimes that support Thread.Abort)
- we could instead add a global using alias, to get the `System.Threading.Monitor`-based code for `net8.0` and lower, but `net9.0` and higher used the `System.Threading.Lock`-based code

e.g.
```CSharp
//GlobalUsings.cs
#if !NET9_0_OR_GREATER
global using Lock = object;
#endif
```

And then update usages
```diff
-private object _lock ...
+private Lock _lock ...
```
